test/app.js

```diff
@@ -0,0 +1,117 @@
+var harvester = require('../lib/harvester');
+var JSONAPI_Error = harvester.JSONAPI_Error;
+var Promise = require('bluebird');
+var config = require('./config.js');
+
+function configureDefaultApp(app) {
+  app.resource('person', {
+    name: String,
+    appearances: Number,
+    pets: ['pet'],
+    soulmate: {ref: 'person', inverse: 'soulmate'},
+    lovers: [
+      {ref: 'person', inverse: 'lovers'}
+    ]
+  })
+
+      .resource('vehicle', {
+        name: String,
+        owners: [
+          {ref: 'person', inverse: 'owners'}
+        ]
+      })
+
+      .resource('pet', {
+        name: String,
+        appearances: Number,
+        owner: 'person'
+      })
+
+      .resource('cat', {
+        name: String
+      }, {namespace: 'animals'})
+
+      .resource('foobar', {
+        foo: String
+      })
+
+      .before(function (req, res) {
+        var foobar = this;
+
+        if (foobar.foo && foobar.foo === 'bar') {
+          // promise
+          return new Promise(function (resolve, reject) {
+            reject(new JSONAPI_Error({
+              status: 400,
+              detail: 'Foo was bar'
+            }));
+          });
+        } else if (foobar.foo && foobar.foo === 'baz') {
+          // non-promise
+          throw new JSONAPI_Error({
+            status: 400,
+            detail: 'Foo was baz'
+          });
+        } else {
+          return foobar;
+        }
+      });
+
+
+  app.router.get('/random-error', function (req, res, next) {
+    next(new Error('this is an error'));
+  });
+
+  app.router.get('/json-errors-error', function (req, res, next) {
+    next(new JSONAPI_Error({status: 400, detail: 'Bar was not foo'}));
+  });
+
+  app.listen(config.harvester.port);
```

Comment:
This'll make it impossible to use this file alongside an events.js process. Right now we're not doing that so it's not really a huge deal, but it matters because other users will look at this file as a demo of best-practices.